File: src/windows-phone-8/NfcPlugin.cs
using System.Runtime.Serialization;
using Windows.Networking.Proximity;
using WPCordovaClassLib.Cordova;
using WPCordovaClassLib.Cordova.Commands;
using WPCordovaClassLib.Cordova.JSON;

using System.Runtime.InteropServices.WindowsRuntime;
using Windows.Storage.Streams;
using System.Diagnostics;
using System.IO;

using System.Collections.Generic;
using System.Text;
using System.Collections;

using ChariotSolutions.NFC.NDEF;

// 
// http://www.nfc-forum.org/specs/spec_list/
// http://msdn.microsoft.com/en-us/library/windows/apps/br241250.aspx
// 

namespace Cordova.Extension.Commands
{
    class NfcPlugin : BaseCommand
    {
        private ProximityDevice proximityDevice;
        private long subscribedMessageId = -1;
        private long publishedMessageId = -1;

        public void init(string args)
        {
            // not used for WP8
        }

        // no args
        public void registerNdef(string args)
        {
            Debug.WriteLine("Registering for NDEF");
            proximityDevice = ProximityDevice.GetDefault();
            subscribedMessageId = proximityDevice.SubscribeForMessage("NDEF", MessageReceivedHandler);
            DispatchCommandResult();
        }

        // no args
        public void removeNdef(string args)
        {
            Debug.WriteLine("Removing NDEF");
            if (subscribedMessageId != -1)
            {
                proximityDevice.StopSubscribingForMessage(subscribedMessageId);
                subscribedMessageId = -1;
            }
            DispatchCommandResult();
        }

        // args[0] is a NdefMessage, which is a JSON array of NdefRecords
        public void writeTag(string args)
        {
            Debug.WriteLine("Write Tag");
            publish("NDEF:WriteTag", args);
            DispatchCommandResult();

            // TODO is there a callback after a successful write so we can stop publishing?
        }

        // args[0] is NdefMessage
        public void shareTag(string args)
        {
            Debug.WriteLine("Share Tag");
            publish("NDEF", args);
            DispatchCommandResult();
        }

        // no args
        public void unshare(string args)
        {
            stopPublishing();
            DispatchCommandResult();
        }

        private void publish(string type, string args)
        {
            string ndefMessage = JsonHelper.Deserialize<string[]>(args)[0];
            NdefRecord[] records = JsonHelper.Deserialize<NdefRecord[]>(ndefMessage);
            byte[] data = NdefMessage.toBytes(records);
            stopPublishing();
            publishedMessageId = proximityDevice.PublishBinaryMessage(type, data.AsBuffer());
        }

        private void stopPublishing()
        {
            if (publishedMessageId != -1)
            {
                proximityDevice.StopPublishingMessage(publishedMessageId);
                publishedMessageId = -1;
            }
        }

        private void MessageReceivedHandler(ProximityDevice sender, ProximityMessage message)
        {

            var bytes = message.Data.ToArray();
            NdefMessage ndefMessage = NdefMessage.parse(bytes);            

            string tag = JsonHelper.Serialize(ndefMessage);

            string[] argsForJavaScriptEvent = new string[] {
                "ndef",
                tag
           };

            // calling a global js method to fire an nfc event
            ScriptCallback script = new ScriptCallback("fireNfcTagEvent", argsForJavaScriptEvent);
            this.InvokeCustomScript(script, false);
        }

    }
}

File: src/windows/nfc-plugin.js
/* globals:  export, module, console, document, ndef, Windows, Uint8Array */

"use strict";

var ndefUtils = {
    // convert Uint8Array to []
    toArray: function (bytes) {
        var output = [], i = 0;
        for (; i < bytes.length; i += 1) {
            output.push(bytes[i]);
        }

        return output;
    }
};

var self = module.exports = {
    init: function (success, failure, args) {
        if (self._initialized) {
            if (success) {
                success();
            }

            return;
        }

        self.subscribedMessageId = -1;
        self.publishedMessageId = -1;
        self.proximityDevice = Windows.Networking.Proximity.ProximityDevice.getDefault();

        if (!self.proximityDevice) {
            console.log("WARNING: proximity device is null");

            if (failure) {
                failure();
            }
        }

        // TODO this never calls success on the first time
        self._initialized = true;
    },
    registerNdef: function (success, failure, args) {
        self.init();

        console.log("Registering for NDEF");

        try {
            self.subscribedMessageId = self.proximityDevice.subscribeForMessage("NDEF", self.messageReceivedHandler);
            success();
        } catch (e) {
            console.log(e);
            failure(e);
        }
    },
    removeNdef: function (success, failure, args) {
        self.init();

        console.log("Removing NDEF");

        try {
            if (self.subscribedMessageId !== -1) {
                self.proximityDevice.stopSubscribingForMessage(self.subscribedMessageId);
                self.subscribedMessageId = -1;
            }

            success();
        } catch (e) {
            console.log(e);
            failure(e);
        }
    },
    writeTag: function (success, failure, args) {
        self.init();

        console.log("Write Tag");

        try {
            var records = args[0];
            var bytes = ndef.encodeMessage(records);

            self.stopPublishing();

            var dataWriter = new Windows.Storage.Streams.DataWriter();
            dataWriter.unicodeEncoding = Windows.Storage.Streams.UnicodeEncoding.utf16LE;
            dataWriter.writeBytes(bytes);

            self.publishedMessageId = self.proximityDevice.publishBinaryMessage("NDEF:WriteTag",
                dataWriter.detachBuffer(),
                function (sender, messageId) {
                      console.log("Successfully wrote message to the NFC tag.");
                      self.stopPublishing();

                      success();
                }
            );

        } catch (e) {
            console.log(e);
            failure(e);
        }
    },
    shareTag: function(success, failure, args) {
        self.init();

        console.log("Share Tag");

        try {
            var records = args[0];
            var bytes = ndef.encodeMessage(records);

            self.stopPublishing();

            var dataWriter = new Windows.Storage.Streams.DataWriter();
            dataWriter.unicodeEncoding = Windows.Storage.Streams.UnicodeEncoding.utf16LE;
            dataWriter.writeBytes(bytes);

            self.publishedMessageId = self.proximityDevice.publishBinaryMessage("NDEF",
                dataWriter.detachBuffer(),
                function (sender, messageId) {
                    console.log("Successfully shared message over peer-to-peer.");
                    self.stopPublishing();

                    success();
                });

        } catch (e) {
            console.log(e);
            failure(e);
        }
    },
    unshareTag: function(success, failure, args) {
        self.init();

        console.log("Unshare Tag");

        try {
            self.stopPublishing();
            success();
        } catch (e) {
            console.log(e);
            failure(e);
        }
    },
    showSettings: function(success, failure, args) {

        // WARNING: this isn't documented, so it might break
        var nfcSettingsUri = "ms-settings-proximity:";
        var uri = new Windows.Foundation.Uri(nfcSettingsUri);

        Windows.System.Launcher.launchUriAsync(uri).then(
            function (settingsAppeared) {
                if (settingsAppeared) {
                    success();
                } else {
                    failure();
                }
            }
        );
    },
    stopPublishing: function() {
        if (self.publishedMessageId !== -1) {
            self.proximityDevice.stopPublishingMessage(self.publishedMessageId);
            self.publishedMessageId = -1;
        }
    },
    messageReceivedHandler: function (sender, message) {
        var bytes = new Uint8Array(message.data.length);
        var dataReader = Windows.Storage.Streams.DataReader.fromBuffer(message.data);
        dataReader.readBytes(bytes);
        dataReader.close();

        var byteArray = ndefUtils.toArray(bytes);
        var ndefMessage = ndef.decodeMessage(byteArray);
        console.log(JSON.stringify(ndefMessage));

        // on windows, tag only contains the ndef message
        // other platforms have tag data
        var tag = {
            ndefMessage: ndefMessage
        };

        // fire JavaScript event with NDEF data
        var e = document.createEvent('Events');
        e.initEvent("ndef", true, false);
        e.tag = tag;
        document.dispatchEvent(e);
    }
}; // exports

require("cordova/exec/proxy").add("NfcPlugin", module.exports);
